File: frontend-api/src/main/java/no/nav/melosys/tjenester/gui/dto/anmodning/AnmodningsperiodeSkrivDto.java
package no.nav.melosys.tjenester.gui.dto.anmodning;

import java.time.LocalDate;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonUnwrapped;
import no.nav.melosys.domain.Anmodningsperiode;
import no.nav.melosys.domain.jpa.LovvalgBestemmelsekonverterer;
import no.nav.melosys.domain.kodeverk.Landkoder;
import no.nav.melosys.domain.kodeverk.LovvalgBestemmelse;
import no.nav.melosys.domain.kodeverk.Trygdedekninger;
import no.nav.melosys.tjenester.gui.dto.periode.PeriodeDto;
import org.apache.commons.lang3.StringUtils;

public class AnmodningsperiodeSkrivDto {
    private static final LovvalgBestemmelsekonverterer konverterer = new LovvalgBestemmelsekonverterer();

    public final String id;
    @JsonUnwrapped(suffix = "Dato")
    public final PeriodeDto periode;
    public final String lovvalgBestemmelse;
    public final String tilleggBestemmelse;
    public final String lovvalgsland;
    public final String unntakFraBestemmelse;
    public final String unntakFraLovvalgsland;
    public final String trygdeDekning;
    public final String medlemskapsperiodeID;


    protected AnmodningsperiodeSkrivDto(String id,
                                        PeriodeDto periode,
                                        LovvalgBestemmelse lovvalgBestemmelse,
                                        LovvalgBestemmelse tilleggBestemmelse,
                                        Landkoder lovvalgsland,
                                        LovvalgBestemmelse unntakFraBestemmelse,
                                        Landkoder unntakFraLovvalgsland,
                                        Trygdedekninger trygdeDekning,
                                        String medlemskapsperiodeID) {
        this.id = id;
        this.periode = periode;
        this.lovvalgBestemmelse = lovvalgBestemmelse != null ? lovvalgBestemmelse.name() : null;
        this.tilleggBestemmelse = tilleggBestemmelse != null ? tilleggBestemmelse.name() : null;
        this.lovvalgsland = lovvalgsland != null ? lovvalgsland.name() : null;
        this.unntakFraBestemmelse = unntakFraBestemmelse != null ? unntakFraBestemmelse.name() : null;
        this.unntakFraLovvalgsland = unntakFraLovvalgsland != null ? unntakFraLovvalgsland.name() : null;
        this.trygdeDekning = trygdeDekning != null ? trygdeDekning.name() : null;
        this.medlemskapsperiodeID = medlemskapsperiodeID;
    }

    @JsonCreator
    @SuppressWarnings("unused")
    public AnmodningsperiodeSkrivDto(Map<String, String> json) {
        this(json.get("id"),
            new PeriodeDto(LocalDate.parse(json.get("fomDato")),
                StringUtils.isEmpty(json.get("tomDato")) ? null : LocalDate.parse(json.get("tomDato"))),
            konverterLovvalgsBestemmelse(json.get("lovvalgBestemmelse")),
            konverterLovvalgsBestemmelse(json.get("tilleggBestemmelse")),
            enumVerdiEllerNull(Landkoder.class, json.get("lovvalgsland")),
            konverterLovvalgsBestemmelse(json.get("unntakFraBestemmelse")),
            enumVerdiEllerNull(Landkoder.class, json.get("unntakFraLovvalgsland")),
            enumVerdiEllerNull(Trygdedekninger.class, json.get("trygdeDekning")),
            json.get("medlemskapsperiodeID"));
    }

    public static AnmodningsperiodeSkrivDto av(Anmodningsperiode anmodningsperiode) {
        return new AnmodningsperiodeSkrivDto(anmodningsperiode.getId().toString(),
            new PeriodeDto(anmodningsperiode.getFom(), anmodningsperiode.getTom()),
            anmodningsperiode.getBestemmelse(),
            anmodningsperiode.getTilleggsbestemmelse(),
            anmodningsperiode.getLovvalgsland(),
            anmodningsperiode.getUnntakFraBestemmelse(),
            anmodningsperiode.getUnntakFraLovvalgsland(),
            anmodningsperiode.getDekning(),
            anmodningsperiode.getMedlPeriodeID() != null ? anmodningsperiode.getMedlPeriodeID().toString() : null);
    }

    public final Anmodningsperiode til() {
        return new Anmodningsperiode(periode.getFom(), periode.getTom(),
            enumVerdiEllerNull(Landkoder.class, lovvalgsland),
            konverterer.convertToEntityAttribute(lovvalgBestemmelse),
            konverterer.convertToEntityAttribute(tilleggBestemmelse),
            enumVerdiEllerNull(Landkoder.class, unntakFraLovvalgsland),
            konverterer.convertToEntityAttribute(unntakFraBestemmelse),
            enumVerdiEllerNull(Trygdedekninger.class, trygdeDekning));
    }

    private static LovvalgBestemmelse konverterLovvalgsBestemmelse(String bestemmelsesnavn) {
        return konverterer.convertToEntityAttribute(bestemmelsesnavn);
    }

    private static <E extends Enum<E>> E enumVerdiEllerNull(Class<E> enumKlasse, String nøkkel) {
        return nøkkel == null ? null : Enum.valueOf(enumKlasse, nøkkel);
    }
}

File: frontend-api/src/main/kotlin/no/nav/melosys/tjenester/gui/dto/anmodning/AnmodningsperiodeSkrivDto.kt
package no.nav.melosys.tjenester.gui.dto.anmodning

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonUnwrapped
import no.nav.melosys.domain.Anmodningsperiode
import no.nav.melosys.domain.jpa.LovvalgBestemmelsekonverterer
import no.nav.melosys.domain.kodeverk.Land_iso2
import no.nav.melosys.domain.kodeverk.LovvalgBestemmelse
import no.nav.melosys.domain.kodeverk.Trygdedekninger
import no.nav.melosys.tjenester.gui.dto.periode.PeriodeDto
import org.apache.commons.lang3.StringUtils
import java.time.LocalDate

open class AnmodningsperiodeSkrivDto(
    val id: String,
    @JsonUnwrapped(suffix = "Dato") val periode: PeriodeDto,
    val lovvalgBestemmelse: LovvalgBestemmelse?,
    val tilleggBestemmelse: LovvalgBestemmelse?,
    val lovvalgsland: Land_iso2?,
    val unntakFraBestemmelse: LovvalgBestemmelse?,
    val unntakFraLovvalgsland: Land_iso2?,
    val trygdeDekning: Trygdedekninger?,
    val medlemskapsperiodeID: String?,
) {
    companion object {
        private val konverterer = LovvalgBestemmelsekonverterer()

        @JvmStatic
        @JsonCreator
        fun fromJson(json: Map<String, String>): AnmodningsperiodeSkrivDto {
            return AnmodningsperiodeSkrivDto(
                json["id"]!!,
                PeriodeDto(
                    LocalDate.parse(json["fomDato"]),
            if (StringUtils.isEmpty(json["tomDato"])) null else LocalDate.parse(json["tomDato"])
                ),
                konverterLovvalgsBestemmelse(json["lovvalgBestemmelse"]),
                konverterLovvalgsBestemmelse(json["tilleggBestemmelse"]),
                enumVerdiEllerNull(Land_iso2::class.java, json["lovvalgsland"]),
                konverterLovvalgsBestemmelse(json["unntakFraBestemmelse"]),
                enumVerdiEllerNull(Land_iso2::class.java, json["unntakFraLovvalgsland"]),
                enumVerdiEllerNull(Trygdedekninger::class.java, json["trygdeDekning"]),
                json["medlemskapsperiodeID"]
            )
        }

        fun av(anmodningsperiode: Anmodningsperiode): AnmodningsperiodeSkrivDto {
            return AnmodningsperiodeSkrivDto(
                anmodningsperiode.id.toString(),
                PeriodeDto(anmodningsperiode.fom, anmodningsperiode.tom),
                anmodningsperiode.bestemmelse,
            anmodningsperiode.tilleggsbestemmelse,
                anmodningsperiode.lovvalgsland,
                anmodningsperiode.unntakFraBestemmelse,
                anmodningsperiode.unntakFraLovvalgsland,
                anmodningsperiode.dekning,
                anmodningsperiode.medlPeriodeID?.toString()
            )
        }

        private fun konverterLovvalgsBestemmelse(bestemmelsesnavn: String?): LovvalgBestemmelse? {
            return konverterer.convertToEntityAttribute(bestemmelsesnavn)
        }

        private fun <E : Enum<E>> enumVerdiEllerNull(enumKlasse: Class<E>, nøkkel: String?): E? {
            return nøkkel?.let {
                java.lang.Enum.valueOf(enumKlasse, it)
            }
        }
    }

    fun til(): Anmodningsperiode {
        return Anmodningsperiode(
            periode.fom,
            periode.tom,
            enumVerdiEllerNull(Land_iso2::class.java, lovvalgsland?.name),
            konverterer.convertToEntityAttribute(lovvalgBestemmelse?.name()),
            konverterer.convertToEntityAttribute(tilleggBestemmelse?.name()),
            enumVerdiEllerNull(Land_iso2::class.java, unntakFraLovvalgsland?.name),
            konverterer.convertToEntityAttribute(unntakFraBestemmelse?.name()),
            enumVerdiEllerNull(Trygdedekninger::class.java, trygdeDekning?.name)
        )
    }
}
